core/services/workflows/syncer/workflow_registry.go

```diff
@@ -415,114 +420,208 @@ func (w *workflowRegistry) syncUsingEventStrategy(ctx context.Context, don capab
+		id := idFor(wfMeta.Owner, wfMeta.WorkflowName)
 		switch {
-		// if the workflow is active, but unable to get engine from the engine registry
-		// then handle as registered event
-		case wfMeta.Status == uint8(WorkflowStatusActive) && engineErr != nil:
-			toRegisteredEvent := WorkflowRegisteredV1{
-				WorkflowID:    wfMeta.WorkflowID,
-				WorkflowOwner: wfMeta.Owner,
-				DonID:         wfMeta.DonID,
-				Status:        wfMeta.Status,
-				WorkflowName:  wfMeta.WorkflowName,
-				BinaryURL:     wfMeta.BinaryURL,
-				ConfigURL:     wfMeta.ConfigURL,
-				SecretsURL:    wfMeta.SecretsURL,
-			}
-			events = append(events, Event{
-				Data:      toRegisteredEvent,
-				EventType: WorkflowRegisteredEvent,
-			})
-			seenMap[engineKeyStr] = true
-
-		// if the workflow is active, the workflow engine is in the engine registry, but the metadata has changed
-		// then handle as updated event
-		case wfMeta.Status == uint8(WorkflowStatusActive) && currWfID != prevWfID:
-			toUpdatedEvent := WorkflowUpdatedV1{
-				OldWorkflowID: engine.WorkflowID,
-				NewWorkflowID: wfMeta.WorkflowID,
-				WorkflowOwner: wfMeta.Owner,
-				DonID:         wfMeta.DonID,
-				WorkflowName:  wfMeta.WorkflowName,
-				BinaryURL:     wfMeta.BinaryURL,
-				ConfigURL:     wfMeta.ConfigURL,
-				SecretsURL:    wfMeta.SecretsURL,
-				Status:        wfMeta.Status,
-			}
-			events = append(events, Event{
-				Data:      toUpdatedEvent,
-				EventType: WorkflowUpdatedEvent,
-			})
-			seenMap[engineKeyStr] = true
-
-		// if the workflow is active, the workflow engine is in the engine registry, and the metadata is the same
-		// then state is properly synced. generate no events.
-		case wfMeta.Status == uint8(WorkflowStatusActive) && currWfID == prevWfID:
-			seenMap[engineKeyStr] = true
+		case wfMeta.Status == WorkflowStatusActive:
+			switch {
+			// if the workflow is active, but unable to get engine from the engine registry
+			// then handle as registered event
+			case !engineFound:
+				signature := fmt.Sprintf("%s-%s-%s", WorkflowRegisteredEvent, currWfID, toSpecStatus(wfMeta.Status))
+
+				if _, ok := pendingEvents[id]; ok && pendingEvents[id].signature == signature {
+					events = append(events, pendingEvents[id])
+					delete(pendingEvents, id)
+					continue
+				}
 
-		// Paused means we skip for processing as a deleted event
-		// To be handled below as a deleted event, which clears the DB workflow spec.
-		case wfMeta.Status == uint8(WorkflowStatusPaused):
-			logger.Debugf("workflow is paused, not generating event")
+				delete(pendingEvents, id)
+
+				toRegisteredEvent := WorkflowRegisteredV1{
+					WorkflowID:    wfMeta.WorkflowID,
+					WorkflowOwner: wfMeta.Owner,
+					DonID:         wfMeta.DonID,
+					Status:        wfMeta.Status,
+					WorkflowName:  wfMeta.WorkflowName,
+					BinaryURL:     wfMeta.BinaryURL,
+					ConfigURL:     wfMeta.ConfigURL,
+					SecretsURL:    wfMeta.SecretsURL,
+				}
+				events = append(events, &reconciliationEvent{
```

Comment:
should this copy over the event count and last try from the pendingEvents?  If not I think it would re-try on every tick where (say) engine creation is failing, thus on each reconcile it would generate new registration event every tick cycle ignoring the fact that registration was already attempted and failed on the previous attempt, so would end up not backing off.  Have I missed something?

Reply:
We do that here: https://github.com/smartcontractkit/chainlink/pull/17475/files/90158f03868a7e2b516ca58a0c6ce60f832bc1f2#diff-488f554caf79f2a49d90f3167297918f99d9c8b76a1e671aa33ec1914863314dR486